src/shogun/lib/ShogunException.h

```diff
@@ -48,5 +48,25 @@ class ShogunException: public std::exception
 		/** exception string */
 		std::string msg;
 };
+
+/** @brief Class InvalidStateException defines an exception which is thrown
+ * whenever an object in Shogun in invalid state is used. For example, a machine
+ * is used before training.
+ */
+class InvalidStateException : public ShogunException
```

Comment:
i would define these in separate headers and cpp files :)

Reply:
and in the case of "a machine is used before training" i would call that exception `MachineNotTrainedException` as invalid state could be a lot of things and the not trained case i believe should be explicitly spelled out for the user :)